### ISSUE TITLE
Obstacles

### ISSUE THREAD

**Issue by MonkeyToiletLadder** [samueltardieu/pathfinding#255]
Is there a way to easily implement obstacles for a_star?

**Comment by samueltardieu** [comment]
This is very easy to do, just don't return positions containing obstacles as neighbours. Or did I misunderstand your question?

**Comment by MonkeyToiletLadder** [comment]
There is no grid structure passed in to a star.  How would i initialize the positions to be obstacles?  Or rather how would i modify astar to allow it to accept a grid of nodes?  I couldnt figure it out from just reading your source code.

**Comment by extrawurst** [comment]
@samueltardieu actually answered it already. the search function do not care how nodes are arranged, it only cares which nodes (`successors`) you can reach from a node: https://docs.rs/pathfinding/2.0.4/pathfinding/directed/astar/fn.astar.html

**Comment by MonkeyToiletLadder** [comment]
Sorry let me rephrase this.  I dont understand how to tell a star what nodes are obstacles.  What data structure do i need to modify to edit the positions to be obstacles?  I know im being a bit needy im sorry.  Maybe i need to reread the code and figure it out for myself, but i would like a point in the right direction.

**Comment by extrawurst** [comment]
The algorithm does not have a concept of obstacles. When returning the reachable `successors` you simply check your own datastructure (in your case the grid) and return the nodes reachable from a node X if one of these nodes should not be reachable (for example because being blocked by an obstacle then you simply do not return it.

The advantage of implementing it like this is that any use case can be applied not only 2d grids but any kind of graph and it’s up to the user to handle that

**Comment by MonkeyToiletLadder** [comment]
Ok i think i get it now.  Thank you.
